src/Bundle.ts

```diff
@@ -162,46 +504,72 @@ export default class Bundle {
 			}
 		});
 
-		this.graph.scope.deshadow(toDeshadow);
+		this.graph.scope.deshadow(toDeshadow, this.orderedModules.map(module => module.scope));
 	}
 
-	private setIdentifierRenderResolutions (options: OutputOptions) {
-		let dynamicImportMechanism: DynamicImportMechanism;
-
-		if (options.format !== 'es') {
-			if (options.format === 'cjs') {
-				dynamicImportMechanism = {
-					left: 'Promise.resolve(require(',
-					right: '))'
-				};
-			} else if (options.format === 'amd') {
-				dynamicImportMechanism = {
-					left: 'new Promise(function (resolve, reject) { require([',
-					right: '], resolve, reject) })'
-				}
+	getModuleDeclarations () {
+		const reexportDeclarations: {
+			[id: string]: ReexportSpecifier[]
+		} = {};
+
+		for (let name in this.exports) {
+			const expt = this.exports[name];
+			// skip local exports
+			if (expt.module.bundle === this)
+				continue;
+			let depId;
+			if (expt.module.isExternal) {
+				depId = expt.module.id;
+			} else {
+				depId = (<Bundle>expt.module.bundle).id;
 			}
+			const exportDeclaration = reexportDeclarations[depId] = reexportDeclarations[depId] || [];
+			exportDeclaration.push({
+				imported: expt.name,
+				reexported: name
+			});
 		}
 
-		this.orderedModules.forEach(module => {
-			if (this.graph.dynamicImport) {
-				module.dynamicImportResolutions.forEach((replacement, index) => {
-					const node = module.dynamicImports[index];
+		const dependencies: BundleDependencies = [];
 
-					if (!replacement)
-						return;
+		let imports: ImportSpecifier[];
+		this.dependencies.forEach(dep => {
```

Comment:
What are you doing with each dependency here? Might be another candidate for extraction (with a good name that answers this question 😉)

Reply:
This function is consolidating the "module declaration" interface for the entire chunk. So it is gathering the export list, and the dependency list in a form that provides all the information about the cross-chunk boundaries. The dependencies loop here is exactly gathering the dependencies part of that interface. Again focus on the output data structures first and foremost over the words of it. Much more descriptive level to look at things.

Reply:
But you have to read a lot of code you are probably not interested in but you cannot be sure you are not interested in to get there. Basically having read this function again what it does is
```javascript
const reexportDeclarations = this.getReexportDeclarations()
const chunkDependencies = getChunkDependencies(reexportDeclarations)
const exports = getChunkExports()
return { dependencies, exports }
```

I know, this might look too simple to the naked eye but it is basically splitting this function into three totally separate, about equally sized concerns with little interdependency. Really easy to find out where I have to look if I want to fine-tune any of these things.

But please keep in mind, these are just my personal suggestions. I would be perfectly happy with keeping the code as it is. I just feel that these kinds of transformations will really help us keep agile as it is really easy to drill down into this kind of code for
* people unfamiliar with it
* people familiar with it that have not looked at it for half a year and probably for forgotten the details